Fixed gateway authentication

attakorns opened the pull request.

attakorns committed:
Fixed gateway authentication

attakorns committed:
Removed , from StartAuthentication() and GwmAuthentication() can frame

chrisstaite commented on vehicle/OVMS.V3/components/vehicle_mgev/src/mg_gwm_auth.cpp:
```diff
@@ -139,43 +141,51 @@ void OvmsVehicleMgEv::GwmAuthentication(canbus* currentBus, uint8_t frameType, u
+            std::stringstream sSeed, sKey;
+            sSeed << std::hex << seed;
+            sKey << std::hex << key;
+            ESP_LOGV(TAG, "GWM auth: seed1 received %s. Replying with key1 %s", sSeed.str().c_str(), sKey.str().c_str());
```
Just use %x rather than taking stringstream.

attakorns replied:
Ok, sorry very rusty with c++ :P will amend.

chrisstaite commented on vehicle/OVMS.V3/components/vehicle_mgev/src/mg_can_handler.cpp:
```diff
@@ -89,7 +89,7 @@ void OvmsVehicleMgEv::IncomingPollFrame(CAN_frame_t* frame)
     }
 
-    uint8_t frameType = frame->data.u8[0] >> 4;
+    uint8_t frameType = frame->data.u8[1] >> 4;
```
I'm not sure that this change is correct. It has much wider effects than the authentication.

attakorns replied:
I was wary of that as well. From my understanding, the first byte is supposed to represent data length right? (The line below it also suggests this: 
`uint8_t frameLength = frame->data.u8[0] & 0x0f;`
)
Are there any other cases where this wouldn't be the case?

chrisstaite replied:
The bottom 4 bits of the first byte are the frame length, the top four bits are the frame type.

attakorns replied:
Oh I see. I think I might have confused frameType with the service ID in the second byte. So in that case, instead of checking `(frameType & 4)` for the gateway and BCM authentication response, we should check `data[0] & 4` instead then?

chrisstaite commented on vehicle/OVMS.V3/components/vehicle_mgev/src/mg_can_handler.cpp:
```diff
@@ -131,14 +131,14 @@ void OvmsVehicleMgEv::IncomingPollFrame(CAN_frame_t* frame)
-        ESP_LOGV(TAG, "Got an gateway authentication response");
-        GwmAuthentication(frame->origin, frameType & 3, data);
+        ESP_LOGV(TAG, "Got a gateway authentication response");
+        GwmAuthentication(frame->origin, frameType & 3, &data[1]);
```
Without this change the frame type above can be accessed.

attakorns replied:
Yep, we can keep `data` here and correct the `GwmAuthentication()` function itself if you'd prefer that.

chrisstaite replied:
It avoids the issues with incompatibility with the rest of the system

attakorns replied:
Ok I'll adjust data within `GwmAuthentication()` itself.

chrisstaite commented:
What car does this fix the authentication on? I heard from testing that the existing implementation worked on the Thai car and had no effect on the EU?

attakorns committed:
Corrected log level

attakorns commented:
> What car does this fix the authentication on? I heard from testing that the existing implementation worked on the Thai car and had no effect on the EU?

It fixed it on my Thai car (I'm Ohm on Slack by the way). I thought it was working before because the car was previously authenticated (using my own test code) then when I switched to your firmware, I saw it managed to read data off the car so I thought everything was well and fine. Later, after revisiting your firmware, with the car not authenticated, I noticed that the gateway authentication would start but nothing would happen after that so I tried logging the CAN bus and noticed that `10 01` was transmitted to the gateway instead of `02 10 01 00 00 00 00 00`. @brickmen also confirmed this in his [log](https://mgevhackers.slack.com/archives/C01EVF3E06M/p1614961170049700).

attakorns committed:
- Corrected frameType in IncomingPollFrame()
- Use serviceId for GWM and BCM authentication instead
- Fixed CAN frame for StartAuthentication() and GwmAuthentication()
- Added VEHICLE_POLL_TYPE_ROUTINECONTROL to vehicle.h

attakorns committed:
Reverted to using ESP_LOGV() in GwmAuthentication()

attakorns commented:
Ok have a look at this version @chrisstaite, have corrected `%x`, `frameType` and `data`. I haven't touched BCM authentication yet though, I'll need to mend that too.

attakorns committed:
Log seed and key in GwmAuthentication() with fixed 8 bits and zero padding

chrisstaite commented on vehicle/OVMS.V3/components/vehicle/vehicle.h:
```diff
@@ -77,47 +77,49 @@ struct DashboardConfig;
 
 
-#define VEHICLE_POLL_TYPE_NONE          0x00
+#define VEHICLE_POLL_TYPE_NONE            0x00
```
Shouldn't need to be modifying anything the base OVMS vehicle.

chrisstaite commented:
Not sure why it keeps getting changed, but charts.js.gz should not be committed. Otherwise, looks good.  Not sure if the mainline will like all the changes to vehicle.h, but we'll see.

attakorns commented:
Thanks @chrisstaite. 

Hmm odd I didn't see charts.js.gz in my VS code source control. I'll keep an eye on it for next time.

Regarding vehicle.h, yeah I'm not sure about that either. If they have a problem with that we can always put the definition for `VEHICLE_POLL_TYPE_ROUTINECONTROL 0x31` in our own MG code.
